hbase-server/src/main/java/org/apache/hadoop/hbase/security/access/SnapshotScannerHDFSAclHelper.java

```diff
@@ -0,0 +1,716 @@
+        .map(path -> new HDFSAclOperation(fs, path, users, operationType, false,
+            HDFSAclOperation.AclType.DEFAULT_ADN_ACCESS))
+        .collect(Collectors.toList());
+    setHDFSAclParallel(hdfsAclOperations).get();
+    // set namespace HDFS acls
+    setNamespaceAcl(Sets.newHashSet(admin.listNamespaces()), users, skipNamespaces, skipTables,
+      operationType);
+  }
+
+  private void setNamespaceAcl(Set<String> namespaces, Set<String> users,
+      Set<String> skipNamespaces, Set<TableName> skipTables,
+      HDFSAclOperation.OperationType operationType)
+      throws ExecutionException, InterruptedException, IOException {
+    Set<String> filterNamespaces =
+        namespaces.stream().filter(ns -> !skipNamespaces.contains(ns)).collect(Collectors.toSet());
+    // set namespace root directories HDFS acls
+    List<HDFSAclOperation> hdfsAclOperations = new ArrayList<>();
+    Set<String> skipTableNamespaces =
+        skipTables.stream().map(t -> t.getNamespaceAsString()).collect(Collectors.toSet());
+    for (String ns : filterNamespaces) {
+      if (operationType == HDFSAclOperation.OperationType.REMOVE
+          && skipTableNamespaces.contains(ns)) {
+        // remove namespace directories default HDFS acls for skip tables
+        hdfsAclOperations.addAll(getNamespaceRootPaths(ns).stream()
+            .map(path -> new HDFSAclOperation(fs, path, users,
+                HDFSAclOperation.OperationType.REMOVE, false, HDFSAclOperation.AclType.DEFAULT))
+            .collect(Collectors.toList()));
+      } else {
+        hdfsAclOperations.addAll(getNamespaceRootPaths(ns).stream()
+            .map(path -> new HDFSAclOperation(fs, path, users, operationType, false,
+                HDFSAclOperation.AclType.DEFAULT_ADN_ACCESS))
+            .collect(Collectors.toList()));
+      }
+    }
+    setHDFSAclParallel(hdfsAclOperations).get();
+    // set table directories HDFS acls
+    Set<TableName> tables = new HashSet<>();
+    for (String namespace : filterNamespaces) {
+      tables.addAll(admin.listTableDescriptorsByNamespace(Bytes.toBytes(namespace)).stream()
+          .map(td -> td.getTableName()).collect(Collectors.toSet()));
+    }
+    setTableAcl(tables, users, skipNamespaces, skipTables, false, operationType);
+  }
+
+  private void setTableAcl(Set<TableName> tableNames, Set<String> users, Set<String> skipNamespaces,
+      Set<TableName> skipTables, boolean setNamespaceAcl,
+      HDFSAclOperation.OperationType operationType)
+      throws ExecutionException, InterruptedException, IOException {
+    Set<TableName> filterTableNames = new HashSet<>();
+    for (TableName tableName : tableNames) {
+      if (!skipTables.contains(tableName)
+          && !skipNamespaces.contains(tableName.getNamespaceAsString())) {
+        filterTableNames.add(tableName);
+      }
+    }
+    List<CompletableFuture<Void>> futures = new ArrayList<>();
+    if (setNamespaceAcl) {
```

Comment:
Only when grant/remove table level ACL,   we need to grant or revoke the namespace ACCESS ACL ? So, I think we should move this  block in upper layer,  I mean don't let the details mess up the abstraction.